### PR TITLE
Limit TextField lengths in runner

### PR DESCRIPTION
**Pull request by berroar**
### What is the context of this PR?
This PR has been raised to add the feature to allow differentiated character limits for TextField answers in runner using the new "max_length" field.

### How to review 
Use the test_textfield schema to test the changes.
Check all tests pass.

### CODE REVIEW

**Comment by berroar** [comment]
[[F] Add feature to limit Runner TextField lengths (M)](https://trello.com/c/u28fMYpl/3621-f-add-feature-to-limit-runner-textfield-lengths-m)

File: app/forms/field_handlers/string_handler.py [hunk 1]
```diff
@@ -1,12 +1,36 @@
...
     MANDATORY_MESSAGE_KEY = "MANDATORY_TEXTFIELD"
 
-    def get_field(self) -> StringField:
-        return StringField(
-            label=self.label, description=self.guidance, validators=self.validators
+    @cached_property
+    def validators(self):
+        validate_with = super().validators
+
+        if self.disable_validation is False:
```
Comment (JamesGardiner) [inline app/forms/field_handlers/string_handler.py]:
```suggestion
        if not self.disable_validation:
```

File: test_schemas/en/test_checkbox_multiple_detail_answers.json [hunk 2]
```diff
@@ -89,7 +89,8 @@
                                                     "type": "TextField",
                                                     "validation": {
                                                         "messages": {
-                                                            "MANDATORY_TEXTFIELD": "Enter your topping choice to continue"
+                                                            "MANDATORY_TEXTFIELD": "Enter your topping choice to continue",
+                                                            "MAX_LENGTH_EXCEEDED": "Your answer is too long, it has to be less than %(max)d characters."
```
[the remark this answers is not in the document]
Comment (berroar) [reply test_schemas/en/test_checkbox_multiple_detail_answers.json]: Have resolved the bug in the detail answer code.

File: app/forms/field_handlers/string_handler.py [hunk 3]
```diff
@@ -1,12 +1,36 @@
...
+    @cached_property
+    def max_length(self):
+        return self.answer_schema.get("max_length", 10000)
+
+    def get_field(self) -> MaxStringField:
+        return MaxStringField(
+            label=self.label,
+            description=self.guidance,
+            validators=self.validators,
+            maxlength=self.max_length,
```
[the remark this answers is not in the document]
Comment (berroar) [reply app/forms/field_handlers/string_handler.py]: Have removed this field type.

File: app/forms/field_handlers/string_handler.py [hunk 4]
```diff
@@ -1,12 +1,36 @@
...
+
+    @cached_property
+    def get_length_validator(self):
+        length_message = self.get_validation_message("MAX_LENGTH_EXCEEDED")
+
+        return validators.length(-1, self.max_length, message=length_message)
+
+    @cached_property
+    def max_length(self):
+        return self.answer_schema.get("max_length", 10000)
```
[the remark this answers is not in the document]
Comment (berroar) [reply app/forms/field_handlers/string_handler.py]: Checked with @donnap573, we are OK to use a default for the max_length.  I have made it consistent with the Textarea handler.

File: app/forms/field_handlers/string_handler.py [hunk 5]
```diff
@@ -1,11 +1,34 @@
-from wtforms import StringField
+from werkzeug.utils import cached_property
+from wtforms import StringField, validators
+
 
 from app.forms.field_handlers.field_handler import FieldHandler
 
 
 class StringHandler(FieldHandler):
+    MAX_LENGTH = 10000
```
Comment (JamesGardiner) [inline app/forms/field_handlers/string_handler.py]:
```suggestion
    MAX_LENGTH = 10_000
```

File: app/forms/field_handlers/string_handler.py [hunk 6]
```diff
@@ -1,11 +1,34 @@
...
+    @cached_property
+    def get_length_validator(self):
+        length_message = self.get_validation_message("MAX_LENGTH_EXCEEDED")
+
+        return validators.length(-1, self.max_length, message=length_message)
+
+    @cached_property
+    def max_length(self):
+        max_length = self.answer_schema.get("max_length", 0)
+        return max_length if max_length > 0 else self.MAX_LENGTH
```
Comment (ajmaddaford) [inline app/forms/field_handlers/string_handler.py]: We could use `self.MAX_LENGTH` in the default for the `get`, something like:

```python
return self.answer_schema.get("max_length", self.MAX_LENGTH)
```

If we're concerned about someone defining a `max_length` of zero or less we should enforce the value to be greater than zero in the json schema.

File: tests/app/forms/field_handlers/test_string_handler.py [hunk 7]
```diff
@@ -21,3 +22,36 @@ class TestForm(Form):
...
+
+    validator = string_handler.get_length_validator
+
+    assert validator.message == "The message is too long!"
+
+
+def test_get_length_validator_with_max_length_override():
+    answer = {"max_length": 30}
+
+    string_handler = StringHandler(answer, {}, AnswerStore(), {})
```
Comment (ajmaddaford) [inline tests/app/forms/field_handlers/test_string_handler.py]: I think the StringHandler class only requires `answer`, so we should be able to drop the empty dicts and AnswerStore from all the places it's instantiated.

**Pull request merged**